blocks/api/parser.js

```diff
@@ -135,54 +135,87 @@ export function getBlockAttributes( blockType, innerHTML, attributes ) {
+
+				// Serialize using the old save
+				contentToValidate = getSaveContent( {
+					...blockType,
+					attributes: migration.attributes,
+					save: migration.save,
+				}, oldAttributes );
+
+				// Migrate the old attributes
+				parsedAttributes = migration.migrate( oldAttributes );
+			}
+		} else {
+			parsedAttributes = getBlockAttributes( blockType.attributes, innerHTML, attributes );
+			contentToValidate = getSaveContent( blockType, parsedAttributes );
+		}
+	} else {
+		shouldFallback = true;
+	}
+
+	// Fallback to the fallback block type
+	if ( shouldFallback ) {
+		// Explicit empty fallback blocks are ignored
+		if ( ! name && ! innerHTML ) {
+			return;
+		}
+
+		const fallbackBlockType = getBlockType( fallbackBlockName );
+		if ( ! fallbackBlockType ) {
+			// eslint-disable-next-line no-console
+			console.warn( `Block ${ name } ignored, no fallback block` );
+			return;
+		}
 
-	// Try finding type for known block name, else fall back again.
-	let blockType = getBlockType( name );
-	const fallbackBlock = getUnknownTypeHandlerName();
-	if ( ! blockType ) {
-		// If detected as a block which is not registered, preserve comment
-		// delimiters in content of unknown type handler.
 		if ( name ) {
-			innerHTML = getCommentDelimitedContent( name, attributes, innerHTML );
+			originalContent = getCommentDelimitedContent( name, attributes, innerHTML, version );
 		}
 
-		name = fallbackBlock;
-		blockType = getBlockType( name );
+		name = fallbackBlockName;
+		parsedAttributes = getBlockAttributes( fallbackBlockType.attributes, originalContent, attributes );
+		contentToValidate = getSaveContent( fallbackBlockType, parsedAttributes );
 	}
 
-	// Include in set only if type were determined.
-	// TODO do we ever expect there to not be an unknown type handler?
-	if ( blockType && ( innerHTML || name !== fallbackBlock ) ) {
-		// TODO allow blocks to opt-in to receiving a tree instead of a string.
-		// Gradually convert all blocks to this new format, then remove the
-		// string serialization.
-		const block = createBlock(
-			name,
-			getBlockAttributes( blockType, innerHTML, attributes )
-		);
-
-		// Validate that the parsed block is valid, meaning that if we were to
-		// reserialize it given the assumed attributes, the markup matches the
-		// original value.
-		block.isValid = isValidBlock( innerHTML, blockType, block.attributes );
-
-		// Preserve original content for future use in case the block is parsed
-		// as invalid, or future serialization attempt results in an error
-		block.originalContent = innerHTML;
-
-		return block;
-	}
+	const block = createBlock(
+		name,
+		parsedAttributes
+	);
+
+	block.isValid = isEquivalentHTML( originalContent, contentToValidate );
+
+	// Preserve original content for future use in case the block is parsed
+	// as invalid, or future serialization attempt results in an error
+	block.originalContent = originalContent;
+
+	return block;
```

Comment:
It kind of looks like this function is operating some kind of pipeline on the raw blocks: get a block from the parser, see if we have version stuff, see if we have fallback stuff, etc…

Since we're getting deep in the nesting do you think it might be time to consider splitting up that logic into isolated and more easily reasoned-about and testable pieces then turn this function into a composition of those steps?

I'm seeing the start of some kind of recursive operation with an accumulator passing things like `[ block, arguments, shouldFallback ]`

```js
if {
  if {
    if {
    } else {
    }
  } else {
  }
} else {
}
```

^^^ that's starting to get hard to follow :wink:

Reply:
Yes, I know, it's not easy to break actually. Just take a look at the flow chart linked above. The "reusable" pieces are already outside the function. Do you have a suggestion on how we can improve it? I'm all for it.